### PR TITLE
fix: Wrong linenumbers reported when linting

### LINKED ISSUE #2574: Wrong error line numbers reported in version 8.0.1
**Snakemake version** 8.0.1

**Describe the bug**
As described in [this issue](https://github.com/snakemake/snakemake/issues/1256), Snakemake reports incorrect line numbers for errors.

**Minimal example**
1. Create the following "Snakefile":
```
rule all:
    input:
        "complete"

rule start:
    output:
        "complete"
    params:
        a = 1    # Edit - this syntax error (the missing comma) is deliberate, generating the erroneous line number in the error message.
        b = 2
    shell:
        "touch complete"
```
2. run snakemake --cores 1

This generates the error:
`SyntaxError in file [path]/Snakefile, line 20:
invalid syntax. Perhaps you forgot a comma? (Snakefile, line 20)`

It is the fact that the error message states "line 20" rather than "line 9" or "line 10" that is the issue I am pointing out.

The file is only 12 lines long and the SyntaxError actually occurs on line 9/10 with the deliberately introduced "forgotten" comma, which is meant to generate the SyntaxError for illustrative purposes.

### PR DESCRIPTION
**Pull request by Kleinjohann**
This PR addresses two issues that led to incorrect linenumbers reported when running `snakemake --lint`.

1. If an error was encountered in `--lint` mode, the stacktrace wasn't cut to remove Snakemake internals and the linenumbers were not adjusted, because the linemap was not available. Fixed by storing the workflow object to make the linemap accessible.
2. For rule lints, the linemap wasn't applied, resulting in wrong linenumbers in the output of `snakemake --lint`. Fixed by applying the linemap before outputting the lints.

I added tests for both cases. 

Fixes #2574. 

### QC

* [X] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [X] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

## Summary by CodeRabbit

- **New Features**
	- Enhanced linting functionality with improved line number retrieval for rules.
	- Introduced a new method to centralize line number logic in the linting process.
	- Added a new test for handling specific error conditions in Snakemake workflows.

- **Bug Fixes**
	- Improved accuracy in asserting reported line numbers for unused parameters during linting tests.

- **Documentation**
	- New configuration file for testing, improving organization and readability of test setups.

### CODE REVIEW

**Review by johanneskoester — APPROVED** [review]

File: tests/tests.py [hunk 1]
```diff
@@ -2048,6 +2048,18 @@
     assert "line 9" in stderr
 
 
+def test_issue2574():
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply tests/tests.py:2051]: Not for tests

**Review by johanneskoester — CHANGES_REQUESTED** [review]
Thanks a lot! Can you address the AI review comments? Also, the code should be formatted with black.

File: tests/tests.py [hunk 2]
```diff
@@ -2048,6 +2048,18 @@ def test_issue1256():
...
+    snakefile = os.path.join(dpath("test_issue2574"), "Snakefile")
+    configfile = os.path.join(dpath("test_issue2574"), "config.yaml")
+    p = subprocess.Popen(f"snakemake -s {snakefile} --configfile {configfile} --lint",
+        shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply tests/tests.py:2055]: Not required.

File: tests/test_linting.py [hunk 3]
```diff
@@ -43,6 +43,9 @@ def test_lint(lint, case):
     except sp.CalledProcessError as e:
         if case == "negative":
             assert e.output.decode().strip()
+            if "not_used_params" in lint.name:
+                # Check that the correct line number is reported
+                assert 'line 3' in e.output.decode().strip()
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply tests/test_linting.py:48]: false positive

File: snakemake/linting/rules.py [hunk 4]
```diff
@@ -8,10 +8,18 @@
...
     def item_desc_json(self, rule):
-        return {"rule": rule.name, "line": rule.lineno, "snakefile": rule.snakefile}
+        lineno = self.get_lineno(rule)
+        return {"rule": rule.name, "line": lineno, "snakefile": rule.snakefile}
+
+    def get_lineno(self, rule):
+        linemaps = self.workflow.linemaps
+        if linemaps and rule.snakefile in linemaps:
+            return linemaps[rule.snakefile][rule.lineno]
+        return rule.lineno
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply snakemake/linting/rules.py:22]: One can also add a type hint for the rule as `rule: Rule` and import the `Rule` from `snakemake.rules`.

[the remark this answers is not in the document]
Comment (Kleinjohann) [reply snakemake/linting/rules.py]: Done, I added annotations for the rule argument and the return type.

File: tests/tests.py [hunk 5]
```diff
@@ -2048,6 +2048,22 @@ def test_issue1256():
...
+        shell=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    stdout, stderr = p.communicate()
+    stderr = stderr.decode()
+    assert p.returncode == 1
+    assert "KeyError" in stderr
+    assert "line 4," in stderr
+
```
[the remark this answers is not in the document]
Comment (Kleinjohann) [reply tests/tests.py]: The `subprocess` module is imported as `sp` at the beginning of the file, I changed this test to refer to it using this abbreviation.

File: snakemake/linting/rules.py [hunk 6]
```diff
@@ -1,17 +1,27 @@
 from itertools import chain
 import re
 import sys
+from typing import Union
 
 from snakemake.io import is_flagged
```
[the remark this answers is not in the document]
Comment (Kleinjohann) [reply snakemake/linting/rules.py]: Done, I removed the unused import.

**Review by johanneskoester — APPROVED** [review]

**Pull request merged**